### ISSUE TITLE
[css-color-4] Computed value and serialization of `Infinity` and `NaN` in color functions

### ISSUE THREAD

**Issue by romainmenke** [w3c/csswg-drafts#8629]
What are the expected values and serializations for these examples?

- `lab(calc(Infinity) 0 0)`
- `lab(50% calc(Infinity) 0)`
- `lab(calc(NaN) 0 0)`
- `lab(50% calc(NaN) 0)`
- ...

_manually setting these values seems illogical but they can be the result of calculations and variables. e.g. `calc(var(--foo) / var(--bar))` where `--bar` is possibly zero._

------

Currently browsers do very different things in some area's.

Chrome and Safari seem to agree that `lab(50% calc(Infinity) 0)` serializes as `lab(50% Infinity 0)`, but that no longer roundtrips because the [required `calc`](https://www.w3.org/TR/css-values-4/#calc-serialize) was removed. That seems like an obvious bug, but maybe I overlooked something.

Chrome renders `lab(50% calc(Infinity) 0)` as white.
Safari renders `lab(50% calc(Infinity) 0)` as black.

_[Chrome also sometimes renders a pink area in certain edge cases](https://bugs.chromium.org/p/chromium/issues/detail?id=1426641)_

WPT doesn't have any tests for these kinds of values.

**Comment by tabatkins** [comment]
> Chrome and Safari seem to agree that lab(50% calc(Infinity) 0) serializes as lab(50% Infinity 0), but that no longer roundtrips because the [required calc](https://www.w3.org/TR/css-values-4/#calc-serialize) was removed.

Yes, that's clearly wrong. The arguments aren't calculations, they're either keywords (channel names) or math functions. We should get a WPT enforcing this.

Rendering is just a consequence of the "clamp it to your actual allowed range" behavior that V&U specifies for infinite (and other arbitrarily-large finite) values. I'm not sure what the actual behavior of a ginormous `a` value should be, but whatever it is, that's the correct rendering. (I suspect white? But I have no actual knowledge here.)

**Comment by romainmenke** [comment]
> I'm not sure what the actual behavior of a ginormous a value should be, but whatever it is, that's the correct rendering. (I suspect white? But I have no actual knowledge here.)

Also unsure.
Consistency would be nice, but I can imagine that this is very sensitive to implementation details.

In our implementation for the `a`/`b` component :
- a very large but finite number: `white`
- positive **and** negative infinite: `black`

[`color.js`](https://colorjs.io/notebook/) seems to produce white and blue after gamut mapping. But I am never sure if I am using that library correctly.

```js
let color = new Color("lab(50% 0 0)");
color.coords[1] = Infinity
color
color.toGamut('srgb')
```

<img width="991" alt="Screenshot 2023-03-22 at 10 26 36" src="https://user-images.githubusercontent.com/11521496/226858802-5c4961d8-822c-470a-bad5-75dc16274db5.png">

**Comment by cdoublev** [comment]
> I have no idea how `NaN` should be handled, so I skipped that for now.

If I am not mistaken, only nested math functions (unresolved before computed value) can produce `NaN`: `calc(min(1em, 0px / 0) + 1px)` would serialize to `calc(1px + min(1em, NaN * 1px))` as a specified value.

  > If a top-level calculation (a math function not nested inside of another math function) would produce a value whose numeric part is `NaN`, it instead act as though the numeric part is `0`.

https://drafts.csswg.org/css-values-4/#top-level-calculation

I think your tests should serialize `Infinity` to lowercase.

  > As usual for CSS keywords, these are ASCII case-insensitive. Thus, `calc(InFiNiTy)` is perfectly valid. However, `NaN` must be serialized with this canonical casing.

https://drafts.csswg.org/css-values-4/#calc-error-constants

Hope it helps. =)

**Comment by romainmenke** [comment]
Thank you @cdoublev

I've changed the tests to lowercase `infinity`, this indeed matches tests under `css-values`.

**Comment by tabatkins** [comment]
The *behavior* of NaN is specified in <https://drafts.csswg.org/css-values/#top-level-calculation> - it gets censored to 0 when it would escape the calculation tree. It serializes as `NaN`, tho, similar to infinity, at whatever level it's been able to infect to, given the simplification rules and what value stage you're serializing.

**Comment by romainmenke** [comment]
For `lab(calc(NaN) 0 0)` or `lab(calc(0 / 0) 0 0)`:
- computed value `lab(0 0 0)`
- serializes as `lab(calc(NaN) 0 0)`

Would that be correct?

**Comment by cdoublev** [comment]
> It serializes as `NaN` [...] at whatever level it's been able to infect to

Thanks for the clarification. My understanding of the spec was that it only serializes as `NaN` in non-top level calculations (or `NaN * 1 unit` when the calculation type is not empty).

  > Would that be correct?

I am not in the position to give an answer but it would say yes, then.

**Comment by romainmenke** [comment]
I've added some test for `NaN` : https://github.com/web-platform-tests/wpt/pull/39137/commits/789c766a0f30486e6f77f869d08f495b1240707b

**Comment by cdoublev** [comment]
> What part of the spec is leading you to believe that a top-level `NaN` doesn't serialize as such? I might have made a mistake.

The part that we both referred to (edited for clarity):

  > - If a top-level calculation would produce a value whose numeric part is `NaN`, it instead act as though the numeric part is `0`. 
  > - If a top-level calculation would produce a value whose numeric part is `0⁻`, it instead acts as though the numeric part is the standard "unsigned" zero.

Taking Example 38 following this part:

  > `calc(-5 * 0)` produces an unsigned zero — the calculation resolves to `0⁻`, but as it’s a top-level calculation, it’s then censored to an unsigned zero.

`calc(0 / 0)` produces `0` — the calculation resolves to `NaN`, but as it’s a top-level calculation, it’s then censored to zero.

Basically my understanding meant that `NaN` could only appear in the serialization of unresolved nested calculations.

---

I am also missing why Chrome currently serializes the next example to `calc(infinity)` instead of `calc(-infinity)`.

  > On the other hand, `calc(1 / calc(-5 * 0))` produces `−∞` [...] the inner calc resolves to `0⁻`, and as it’s not a top-level calculation, it passes it up unchanged to the outer calc to produce `−∞`

**Comment by tabatkins** [comment]
> The part that we both referred to (edited for clarity):

That's about how the value actually acts in the property it's used in, when the final result is a NaN. The simplification and serialization of the calculation itself is well-defined by the simplification and serialization algos, and they preserve NaNs at whatever level they show up in.

> I am also missing why Chrome currently serializes the next example

Looks like a bug.

**Comment by romainmenke** [comment]
The tests were merged : https://github.com/web-platform-tests/wpt/pull/39137

------

That only leaves the open question of how to handle enormous values :

> Rendering is just a consequence of the "clamp it to your actual allowed range" behavior that V&U specifies for infinite (and other arbitrarily-large finite) values. I'm not sure what the actual behavior of a ginormous a value should be, but whatever it is, that's the correct rendering. (I suspect white? But I have no actual knowledge here.)

I suspect the right value will just fall out of implementations that are free of bugs but maybe good to have a few examples/tests?

@svgeesus Do you know or have an opinion on how something like `lab(50% calc(Infinity) 0)` which has an absurdly large `a` should behave?

_Anyone writing `lab(50% calc(Infinity) 0)` manually should expect bogus results, but `infinity` can be the result of a calculation._

**Comment by svgeesus** [comment]
> @svgeesus Do you know or have an opinion on how something like `lab(50% calc(Infinity) 0)` which has an absurdly large `a` should behave?

All colors of the form `lab(50% calc((var(--n)*100) 0)` where `n` > 0 have the same (LCH) hue, and increasing chroma. Thus, even if the color itself is imaginary, the gamut mapped used value of the color should be `lab(50% x 0)` where `x` is the greatest chroma that can be displayed for that hue and lightness. I would apply the same logic to +Infinity (and -Infinity gives the opposite hue).

Where `b` is not zero, as `a` increases the contribution of `b` on the resulting hue decreases, so as `a` tends to infinity the hue tends towards `b` being effectively zero. So I guess infinite values of  either `a` or `b` make the other component powerless? And thus there are _eight_ effective hues that the color tends towards, depending on the sign of the component and whether only one, or both, have Infinity values?

**Comment by romainmenke** [comment]
> And thus there are eight effective hues that the color tends towards, depending on the sign of the component and whether only one, or both, have Infinity values?

This is specific to `lab` and `oklab` right?

For `rgb` or `xyz` color spaces an increase in value for any channel also has an effect on lightness, so any enormous value forces the color to black or white depending on the sign.


Interesting, not what I would have guessed but it makes a lot of sense!

**Comment by mysteryDate** [comment]
> For `lab(calc(NaN) 0 0)` or `lab(calc(0 / 0) 0 0)`:
> 
> * computed value `lab(0 0 0)`
> * serializes as `lab(calc(NaN) 0 0)`
> 
> Would that be correct?

@tabatkins Even if this is correct, per spec, why do we _want_ this behavior? Does it have significant use-cases? Do other CSS properties perform this way? The reason I ask is because this would invalidate a couple major assumptions that chromium's color parser makes. Namely that the result of parsing a color function will either be failure or a `Color`, i.e. something that can be encoded with a colorspace and 4 values, which are numbers or "None." 

Tests verifying this are part of interop 2023 and all UAs are currently failing:

https://wpt.fyi/results/css/css-color/parsing/color-valid-lab.html

To pass these tests will involve restructuring things quite a bit and creating a sort of "unresolved color" type so that the color can be serialized differently at parse time than at computed time. Obviously, none of this is impossible, but I simply wanted to highlight that it is a non-trivial amount of work to support. If it's a minor use-case and other UAs are uninterested in supporting it, than I'd propose changing the spec such that:

For `lab(calc(NaN) 0 0)` or `lab(calc(0 / 0) 0 0)`:
 
 * computed value `lab(0 0 0)`
 * serializes as `lab(0 0 0)`

And for `lab(0 calc(infinity) 0)`:
 
 * computed value `lab(0 Infinity 0)`
 * serializes as `lab(0 Infinity 0)`
 
 So, importantly, dropping the "calc" so that the channels are all numbers. Why is it important that they roundtrip in this context?

**Comment by romainmenke** [comment]
> So, importantly, dropping the "calc" so that the channels are all numbers. Why is it important that they roundtrip in this context?

Part of the reason for initially opening this issue is that dropping `calc` "breaks" the value.

- `lab(0 Infinity 0)` -> invalid
- `lab(0 calc(Infinity) 0)` -> something very pink-ish in Chrome

Serializing into an invalid value seems like a bug to me.

**Comment by mysteryDate** [comment]
But maybe `lab(0 calc(Infinity) 0) ` _should_ be invalid, if `lab(0 Infinity 0)` is invalid? Adding a `calc` to a single term to make something valid seems like a bug to _me_.

**Comment by romainmenke** [comment]
> Adding a calc to a single term to make something valid seems like a bug to me.

This is not to make something valid, this is done to express `NaN` or `Infinity` : https://drafts.csswg.org/css-values-4/#calc-serialize

There will be other examples that don't have anything to do with color :)

**Comment by mysteryDate** [comment]
Ooooooooooh, okay. I understand that now. So `color(srgb 0 calc(Infinity) 0)` represents a color with an infinite green color channel, whereas `color(srgb 0 Infinity 0)` is literally an invalid string to represent a color?

This makes sense. Now my next question would be, why are we forcing valid colors out of these inputs? Wouldn't the least surprising outcome for the user be to reject them entirely as colors?

**Comment by mysteryDate** [comment]
Agenda+

**Comment by svgeesus** [comment]
>  I think it would be best if these were all invalid colors

I certainly can't think of an actual use case for such colors.

**Comment by mysteryDate** [comment]
If we decide these are valid colors, then as far as I understand it:
- At parse time: all things equivalent to `NaN` are serialized as `calc(NaN)`, and infinity is serialized as `calc(infinity)` or `calc(-infinity)`
- At computed value time: all `NaN`s become zero. Infinity remains infinity unless it expressing a clamped value, in which case it becomes the maximum possible value in that range. Negative infinity does the same, going to the minimum possible value.

Is that, in general, correct? If so, I'll start adding more test cases to wpt.

**Comment by tabatkins** [comment]
> At parse time:

More specifically, *all* math functions serialize as some variety of math function when they're specified values, regardless of what's inside them. If they've been able to simplify down to a single value, they'll serialize with a `calc()` around them. (Browsers have, generally, not done this correctly in many places, colors being a notable example.)

> At computed value time:

Yup, correct. (And note, just for completeness, that the `infinity` keywords have to still be serialized with a `calc()` wrapper, because they're only valid as keywords *inside a calculation*.)

**Comment by cdoublev** [comment]
This comment above says...

  > [...] the "clamp it to your actual allowed range" behavior that V&U specifies for infinite (and other arbitrarily-large finite) values.

... like this [code comment on WPT](https://github.com/web-platform-tests/wpt/blob/abcbb64783af5b776a3c5f4e13cf4d8756217830/css/css-color/parsing/color-computed-rgb.html#L94C8-L94C112):

  > `calc(infinity)` resolves to the upper bound while `calc(-infinity)` and `calc(NaN)` resolves the lower bound.

But this is not what I understand from what [CSS Values and Units](https://drafts.csswg.org/css-values-4/#calc-range):

  > [...] the value resulting from a top-level calculation must be clamped to the range allowed in the target context.

Out of range channel values are *allowed* in all color functions.

Therefore I think it would be great to clarify when and how `calc(infinity)` resolves when serializing (non-mixed) `hsl()` and `hwb()` as components of a declared value, ie. either before or after color space conversion.

**Comment by svgeesus** [comment]
Re-opening because arguing on the basis of one lagacy function `rgba()` which happens to be clamped, was not sufficient.

**Comment by svgeesus** [comment]
Solved by resolution of 

 - https://github.com/w3c/csswg-drafts/issues/10305

**Comment by svgeesus** [comment]
The CSS Working Group just [discussed](https://www.w3.org/2025/08/19-css-color-minutes.html#2a38) `[css-color-4] Computed value and serialization of Infinity and NaN in color functions` and agreed to the following:

 * `RESOLUTION: Close issue as (now) invalid`

<details><summary>The full IRC log of that discussion</summary>
TabAtkins: because we have to resolve things and don't resolve calc(), naked infinity is invalid, have to wrap in calc() <br>
TabAtkins: no longer necessary, and at used value time you get a large value that we clamp to <br>
romain: +1 <br>
TabAtkins: objections to close as invalid (now)? <br>
matthieud: so infinity resolves as infinity? <br>
TabAtkins: computed would be calc(infinity), and used value is the clamped used value <br>
RESOLUTION: Close issue as (now) invalid <br>
</details>
